Improve Q vector generation mechanism

MBartkowiakSTFC opened the pull request:
**Description of work**
Includes off-diagonal components of HKL indices when finding the extent of HKL values to be used in vector generation.

closes #974

**Fixes**
1. `SphericalLatticeQVectors` uses all vector components when checking the HKL range to cover.
2. Q vector generators calculate HKL values of vectors if a unit cell has been provided, even if the vector generator does not use the unit cell definition to generate vectors.
3. `IQVectors.create()` now takes a `UnitCell` instead of a `Configuration` as optional input.
4. Unit tests have been added to check if spherical vectors are generated the same for equivalent unit cell definitions.

**To test**
Unit tests should all pass.

MBartkowiakSTFC committed:
Add a failing q vector test

Correct vector generation for spherical lattice

Group tests based on expected vector symmetry

Group tests based on expected vector symmetry

Include y=+/-1 in points on sphere

Generate all valid points in a sphere

Generate points on a sphere in SphericalLatticeQVectors

Remove 0,0,0 from q-vector list

Run IJob.check_status before IJob.initialize

Generate Q vectors on a sphere

Fill reciprocal space continuously with vectors

Add probability of vectors being included

Correct probability for q^2 sphere surface

Sample SphericalLatticeQVectors Ising-like

Write out q_vector counts as attribute

Write out q_vector counts as attribute

Make DispersionQVectors independent of unit cell

Use the same sampling in linear, circular, spherical generators

Update docstrings and documentation pages

Split directional vector unit tests

Fix Geometry.py after merging changes

Implement changes from code review

Adjust unit tests to new sphere points function

Add weights attribute to all vector generators

Make all tests pass

Update the vector statistics preview

Add 3D vector preview

Add angle distribution to vectors preview

Add angle distribution to vectors preview

Combine two vector plots into one

Prevent infinite recursion in vector preview

Enable vector statistics on shells in files

Fix wrong dispersion lattice vectors

Adjust q vector weights in analysis runs

Improve docstrings and test results

Relax the tolerance in test_dcsf

Implement changes from code review

Apply suggestions from code review

Co-authored-by: Jacob Wilkins <46597752+oerc0122@users.noreply.github.com>
Signed-off-by: Maciej Bartkowiak <108934199+MBartkowiakSTFC@users.noreply.github.com>

ChiCheng45 commented on MDANSE/Src/MDANSE/Framework/QVectors/SphericalLatticeQVectors.py:
```diff
@@ -61,12 +61,14 @@ def _generate(self):
+        hklMax = np.max(
+            np.ceil(self.qvectors_to_hkl(qMax * np.eye(3), self._unit_cell)) + 1, axis=0
+        )
 
         hkl_vects = np.mgrid[
-            -hklMax[0, 0] : hklMax[0, 0] + 1,
-            -hklMax[1, 1] : hklMax[1, 1] + 1,
-            -hklMax[2, 2] : hklMax[2, 2] + 1,
+            -hklMax[0] : hklMax[0] + 1,
+            -hklMax[1] : hklMax[1] + 1,
+            -hklMax[2] : hklMax[2] + 1,
```
I think this still might miss some q vectors out. Here is a poorly drawn diagram.

<img width="358" height="249" alt="Image" src="https://github.com/user-attachments/assets/500fdab1-ee40-4fda-80a8-013fbde80459" />

Some q-vectors near qmax will be missed, circled in purple In the diagram. The smaller the angle between the reciprocal lattice vectors, the more q-vectors you'd miss out.

MBartkowiakSTFC replied:
I think you're right. I'll make further changes to make sure it works.

MBartkowiakSTFC commented:
The new mechanism of Q vector generation in `SphericalLatticeQVectors` is this:
1. For each |q|, generate a sphere of vectors with length |q|.
2. Convert all vectors into HKL coordinates (fractional so far) and round them to the nearest integer values.
3. Convert HKL values back to lab frame q vectors, and discard those that are now outside of the `width` tolerance.
4. Randomly pick the requested number of vectors from the remaining q vectors.

One problem with this approach is that increasing the `width` does not necessarily increase the spread of |q|, since `width` is used as tolerance cutoff. The advantage is that we stay as close to spherical distribution of vectors as the unit cell definition allows.

ChiCheng45 commented:
> The new mechanism of Q vector generation in `SphericalLatticeQVectors` is this:
> 
> 1. For each |q|, generate a sphere of vectors with length |q|.
> 2. Convert all vectors into HKL coordinates (fractional so far) and round them to the nearest integer values.
> 3. Convert HKL values back to lab frame q vectors, and discard those that are now outside of the `width` tolerance.
> 4. Randomly pick the requested number of vectors from the remaining q vectors.
> 
> One problem with this approach is that increasing the `width` does not necessarily increase the spread of |q|, since `width` is used as tolerance cutoff. The advantage is that we stay as close to spherical distribution of vectors as the unit cell definition allows.

Looks good, but are we sure we want to do it this way? This will mean that the width parameter will be different for SphericalLatticeQVectors from all other qvector generators.

MBartkowiakSTFC commented:
I'm happy to think about it some more. I could probably check the standard deviation of |q| in a shell, and extent the shell if the spread of the vector length was not large enough. I guess that our users may want to simulate the effects of low q resolution on the results, and then they should get the |q| as smeared as they requested.

MBartkowiakSTFC commented:
This PR may be easier to test once #990 has been approved and merged, since we will have the tools to visualise the distribution of vector lengths in individual shells.

ChiCheng45 commented:
> The new mechanism of Q vector generation in `SphericalLatticeQVectors` is this:
> 
> 1. For each |q|, generate a sphere of vectors with length |q|.
> 2. Convert all vectors into HKL coordinates (fractional so far) and round them to the nearest integer values.
> 3. Convert HKL values back to lab frame q vectors, and discard those that are now outside of the `width` tolerance.
> 4. Randomly pick the requested number of vectors from the remaining q vectors.
> 
> One problem with this approach is that increasing the `width` does not necessarily increase the spread of |q|, since `width` is used as tolerance cutoff. The advantage is that we stay as close to spherical distribution of vectors as the unit cell definition allows.

I was thinking that if we adjust it slightly and do this

1. For each |q|, generate a shell of vectors with length |q| +/- width/2 similarly to SphericalQVectors.
2. Convert all vectors into HKL coordinates (fractional so far) and round them to the nearest integer values.
3. Convert HKL values back to lab frame q vectors.

we might be able to solve the weighting issue seen in https://github.com/ISISNeutronMuon/MDANSE/pull/990#issuecomment-3333945282 and #806. Here is the reasoning: say we have a lattice with the following population,

<img width="727" height="424" alt="image" src="https://github.com/user-attachments/assets/9b9d56bb-9144-42e8-8e5f-b5886ec89b90" />

and if we generate uniformly q-vectors in the shell and then snap them to the reciprocal lattice vectors, we should get a similar number of q-vectors going to each |q| bin from the above. For the red shell the first bin would get more per reciprocal lattice vector. The same argument applies if we have more reciprocal lattice vectors at specific angles, so that it also fixes #806. Maybe it might also be worth updating the q-vector visualizer to plot the distribution in terms of spherical coordinates?

The main problem will be that there will be a bit of spilling of some q-vectors out of the range |q| +/- width/2 but it might be fine.

ChiCheng45 commented on MDANSE/Src/MDANSE/Framework/QVectors/SphericalQVectors.py:
```diff
@@ -53,25 +54,53 @@ def _generate(self):
+                    radii * np.sin(theta) * np.sin(phi),
+                    radii * np.cos(theta),
+                )
+            )
```
If I understand this correctly, you are trying to generate a uniform sample on the surface of a sphere. Sampling uniformly across theta and phi will mean that you'll get points bunched up near the poles.

Here is a link that shows the case when theta and phi are taken from a uniform random distribution.
https://mathworld.wolfram.com/SpherePointPicking.html

ChiCheng45 commented on MDANSE/Src/MDANSE/Mathematics/Geometry.py:
```diff
@@ -110,51 +115,63 @@ def moment_of_inertia(
+        number of points on the sphere, by default 1000
 
-def random_points_on_sphere(radius=1.0, nPoints=100):
-    points = np.zeros((3, nPoints), dtype=np.float64)
+    Returns
+    -------
+    npt.NDArray[float]
+        (n_samples, 3) array of point coordinates on a sphere
+    """
 
-    theta = 2.0 * np.pi * np.random.uniform(nPoints)
-    u = np.random.uniform(-1.0, 1.0, nPoints)
-    points[0, :] = radius * np.sqrt(1 - u**2) * np.cos(theta)
-    points[1, :] = radius * np.sqrt(1 - u**2) * np.sin(theta)
-    points[2, :] = radius * u
+    indices = np.arange(n_samples)
+    angle_step = np.pi * (np.sqrt(5.0) - 1.0)  # angle step in radians
 
-    return points
+    ys = 1 - (indices / (n_samples - 1)) * 2  # y coordinate in [-1, 1] range
+    radius = np.sqrt(1 - ys**2)  # circle radius at height y
 
+    theta = angle_step * indices
 
-def random_points_on_circle(axis, radius=1.0, nPoints=100):
-    axis = Vector(axis).normal().array
+    xs = np.cos(theta) * radius
+    zs = np.sin(theta) * radius
 
-    points = np.random.uniform(-radius, radius, 3 * nPoints)
-    points = points.reshape((3, nPoints))
+    return np.vstack([xs, ys, zs])
```
I know that this was in MDANSE originally, but I wonder what the reason was for using this method to generate points on a sphere. Wasn't MDANSE supposed to generate q-vectors randomly rather than this non-random method?

ChiCheng45 commented on MDANSE/Src/MDANSE/Mathematics/Geometry.py:
```diff
@@ -110,51 +115,63 @@ def moment_of_inertia(
+            [2 * (bc - ad), aa + cc - bb - dd, 2 * (cd + ab)],
+            [2 * (bd + ac), 2 * (cd - ab), aa + dd - bb - cc],
+        ]
+    )
```
I think this can be replaced with scipy. 

https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.html

I think you can use from_rotvec() and then as_matrix().

lucas-wilkins replied:
I've found those classes to be quite heavyweight, I can't remember what I was doing, but using scipy's rotation classes made it really slow.

ChiCheng45 replied:
Yea I have seen that too. Actually It looks like `def rotation_matrix(...)` is never used maybe we can just remove it?

MBartkowiakSTFC replied:
OK, `rotation_matrix` is not used and can be removed.
At the same time, `circle_rotation_matrix` is still used, uses the same scipy classes and may have the same performance issues.
Ultimately, we should be careful with it, but I'm pretty sure that we only call this function once per run, so it is not going to be critical to the performance.

ChiCheng45 commented:
One more thing I just realized that we need to thing about.

<img width="764" height="443" alt="image" src="https://github.com/user-attachments/assets/390a691d-9fb3-4637-a264-c3e610531e0d" />

I think we might want to adjust the q-vector so instead we get a certain amount of q-vectors per shell volume. Note that we currently use a fixed number of vectors for each shells so for the shells with the largest |q| will not be sampled as well as the smaller ones. We adjust need to adjust step 1 from the above so that the q-vector generator to sampled a sphere uniformly.

MBartkowiakSTFC commented:
> I think we might want to adjust the q-vector so instead we get a certain amount of q-vectors per shell volume. Note that we currently use a fixed number of vectors for each shells so for the shells with the largest |q| will not be sampled as well as the smaller ones. We adjust need to adjust step 1 from the above so that the q-vector generator to sampled a sphere uniformly.

We can think of doing this, but I am not sure if the larger shells really need this level of fine sampling. Probably some convergence testing would help to see if the additional points bring in new information.

If one day we wanted to add error bars to the results, increasing the sampling should reduce the errors even if the results are largely the same.

ChiCheng45 commented:
> > I think we might want to adjust the q-vector so instead we get a certain amount of q-vectors per shell volume. Note that we currently use a fixed number of vectors for each shells so for the shells with the largest |q| will not be sampled as well as the smaller ones. We adjust need to adjust step 1 from the above so that the q-vector generator to sampled a sphere uniformly.
> 
> We can think of doing this, but I am not sure if the larger shells really need this level of fine sampling. Probably some convergence testing would help to see if the additional points bring in new information.
> 
> If one day we wanted to add error bars to the results, increasing the sampling should reduce the errors even if the results are largely the same.

It depends how you set it. We currently could also be sampling the smaller shells we more q-vectors than is needed.

MBartkowiakSTFC commented:
Changes since the last time:
1. Vector preview now has two tabs: the statistics of |q| and the vector coordinate distribution of specific shells.
2. Plot creator can create the |q| statistics when a file containing vectors is right-clicked. It can also create the plot of vector coordinates for a specific shell when a shell dataset is right-clicked.
3. Vectors are generated together with weights. For most generators the weight of each vector is 1. For lattice generators, vectors are first randomly generated, and then rounded to the nearest integer value of HKL. If multiple vectors ended up at the same HKL, the number of the initial vectors corresponding to this HKL value is the weights of the vector.
4. DISF, DCSF, EISF and CCF calculations now include vector weights in the calculations. Typically, the `rho` values for each vector are multiplied by `sqrt(weight)`, and the resulting correlation for each vector ends up being scaled by `weight`.

oerc0122 commented on Doc/docstrings_to_rst.py:
```diff
@@ -1,22 +1,32 @@
 configurators = sorted(IConfigurator.indirect_subclasses())
 converters = sorted(Converter.indirect_subclasses())
 jobs = sorted(IJob.indirect_subclasses())
+generators = sorted(classname for classname in IQVectors.indirect_subclasses() if classname not in ("IQVectors", "LatticeQVectors",))
```
```suggestion
generators = sorted(set(IQVectors.indirect_subclasses()) - {"IQVectors", "LatticeQVectors"})
```

oerc0122 replied:
Can remove `set()` if `IQVectors.indirect_subclasses` already returns one.

oerc0122 commented on Doc/docstrings_to_rst.py:
```diff
@@ -76,6 +86,38 @@ def make_configurator_doc(conf: str, parent: str) -> str:
+    result += f"{qvecs}\n"
+    result += "".join(len(qvecs)*['~'])+'\n\n'
+    if temp.__doc__:
+        result += '\n'.join(str(x).lstrip() for x in temp.__doc__.split('\n'))
```
Is [`textwrap.dedent`](https://docs.python.org/3/library/textwrap.html#textwrap.dedent) what you want here?

If not could be simpler (since `.split` implies it's a string anyway):

```suggestion
        result += "\n".join(map(str.lstrip, temp.__doc__.splitlines()))
```

oerc0122 commented on Doc/docstrings_to_rst.py:
```diff
@@ -76,6 +86,38 @@ def make_configurator_doc(conf: str, parent: str) -> str:
+        continue
+    result += f"\n.. _qvectors-reference-{qvecs}:\n\n"
+    result += f"{qvecs}\n"
+    result += "".join(len(qvecs)*['~'])+'\n\n'
```
```suggestion
    result += "~" * len(qvecs) + "\n\n"
```

oerc0122 commented on Doc/docstrings_to_rst.py:
```diff
@@ -76,6 +86,38 @@ def make_configurator_doc(conf: str, parent: str) -> str:
+        result += '\n'.join(str(x).lstrip() for x in temp.__doc__.split('\n'))
+    else:
+        print(f"bad docstring in {conf}")
+    if not result.endswith('\n'):
+        result += '\n'
```
This could just be stuck on the end of the `join` above. No other path doesn't end with `\n`

oerc0122 commented on Doc/docstrings_to_rst.py:
```diff
@@ -76,6 +86,38 @@ def make_configurator_doc(conf: str, parent: str) -> str:
+    result += "\nInputs:\n\n"
+    for iname, itype in temp.settings.items():
+        conf = itype[0]
+        defval = itype[1].get("default", "N/A")
```
```suggestion
    for iname, (conf, params) in temp.settings.items():
        defval = params.get("default", "N/A")
```

oerc0122 commented on Doc/docstrings_to_rst.py:
```diff
@@ -76,6 +86,38 @@ def make_configurator_doc(conf: str, parent: str) -> str:
+    target.write("List of q-vector generators\n")
+    target.write("===========================\n\n")
+    for entry in qvector_page:
+        target.write(entry + "\n")
```
```suggestion
with open("pages/vector_generators.rst", "w") as target:
    target.write("""\
.. _vector-generator-list:

List of q-vector generators
===========================
""")
    target.write("\n".join(qvector_page) + "\n")
```

or if you prefer:
```suggestion
with open("pages/vector_generators.rst", "w") as target:
    target.write(
        textwrap.dedent(
            """\
            .. _vector-generator-list:
            
            List of q-vector generators
            ===========================
            """
        )
    )
    target.write("\n".join(qvector_page) + "\n")
```

oerc0122 commented on Doc/docstrings_to_rst.py:
```diff
@@ -76,6 +86,38 @@ def make_configurator_doc(conf: str, parent: str) -> str:
+
 with open("pages/converters.rst", 'w') as target:
     target.write(".. _converter-list:\n")
     target.write("\n")
```
```suggestion
with open("pages/converters.rst", "w") as target:
    target.write(".. _converter-list:\n\n")
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/CurrentCorrelationFunction.py:
```diff
@@ -438,7 +440,7 @@ def run_step(self, index: int):
                         "ik,ij->ikj",
                         veloc,
-                        np.exp(1j * temp_dotprod),
+                        np.exp(1j * temp_dotprod) * np.sqrt(qvec_weights[None, :]),
```
Does `np.sqrt(qvec_weights[None, :])` want to become a pre-computed variable? Might be expensive?

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/SolventAccessibleSurface.py:
```diff
@@ -445,7 +445,7 @@ def initialize(self):
             generate_sphere_points(self.configuration["n_sphere_points"]["value"]),
             dtype=np.float64,
-        )
+        ).T
```
Curious as to why? Should this be it's own PR?

MBartkowiakSTFC replied:
This PR changed the `generate_sphere_points` function which is used by q vector generators, but also by `SolventAccessibleSurface`. This change was needed here to keep the SAS calculation working correctly after the changes to q-vector generators have been made.

oerc0122 commented on MDANSE/Src/MDANSE/Framework/OutputVariables/IOutputVariable.py:
```diff
@@ -122,6 +123,8 @@ def __new__(
 
         obj.scaling_factor = 1.0
 
+        obj.custom_field = custom_field if custom_field else []
```
Is this going to be changed/updated/appended to later? Does it want to be a tuple instead?

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/CircularQVectors.py:
```diff
@@ -18,16 +18,94 @@
+    npt.NDArray[float]
+        a 3x3 rotation matrix
+    """
+    initial_circle_axis = np.array([0, 0, 1])
```
Would scipy be better used here?

https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.align_vectors.html#scipy.spatial.transform.Rotation.align_vectors

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/CircularQVectors.py:
```diff
@@ -18,16 +18,94 @@
+        (
+            np.sin(parameter_points),
+            np.cos(parameter_points),
+            np.zeros(len(parameter_points)),
```
```suggestion
            np.zeros_like(parameter_points),
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/CircularQVectors.py:
```diff
@@ -18,16 +18,94 @@
+            temp_radii[
+                np.where(np.logical_and((temp_radii >= qmin), (temp_radii <= qmax)))
+            ]
+        )
```
Seems like what you want here is a [`truncnorm`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.truncnorm.html)

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/DispersionLatticeQVectors.py:
```diff
@@ -20,52 +20,77 @@
+    generated as :math:`\mathbf{k}_{start} + n\mathbf{k}_{step}` for
+    integer n from 0 to n_steps.
+
+    The 'using_supercell' input is only used for the final conversion
```
Do you think non-diagonal supercells might be used? They are for a lot of phonon calculations.

MBartkowiakSTFC replied:
I admit I did not think that a non-diagonal supercell would be used for an MD simulation. But I guess that anything is possible. Do you think I should extend this input to be a 3x3 array?

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/GridQVectors.py:
```diff
@@ -91,6 +93,7 @@ def _generate(self):
                 "q": q,
                 "q_vectors": vects[:, v],
                 "n_q_vectors": len(v),
+                "weights": np.ones(len(v)),
```
```suggestion
                "weights": np.ones_like(v),
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/IQVectors.py:
```diff
@@ -106,6 +107,132 @@ def hkl_to_qvectors(self, hkls: np.array, unit_cell: UnitCell) -> np.ndarray:
+        scaling /= np.max(scaling)
+        while len(subset) < nvecs_per_shell:
+            index = np.unique(np.random.randint(0, len(q_lengths), nvecs_per_shell))
+            index = index[np.where(np.logical_not(np.in1d(index, subset)))]
```
Should be [np.isin](https://numpy.org/doc/stable/reference/generated/numpy.isin.html#numpy.isin) if we're migrating to 2+ which we should really..?

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/IQVectors.py:
```diff
@@ -106,6 +107,132 @@ def hkl_to_qvectors(self, hkls: np.array, unit_cell: UnitCell) -> np.ndarray:
+        while len(subset) < nvecs_per_shell:
+            index = np.unique(np.random.randint(0, len(q_lengths), nvecs_per_shell))
+            index = index[np.where(np.logical_not(np.in1d(index, subset)))]
+            rand_pool = np.random.ranf(len(index))
```
Also to be deprecated, should be using [`random`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.random.html#numpy.random.Generator.random) from a generator.

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/IQVectors.py:
```diff
@@ -106,6 +107,132 @@ def hkl_to_qvectors(self, hkls: np.array, unit_cell: UnitCell) -> np.ndarray:
+        # spherical_correction = 1 / q_lengths**2
+        scaling = distribution  # * spherical_correction
+        scaling /= np.max(scaling)
+        while len(subset) < nvecs_per_shell:
```
Really, I think this probably wants to be a generator yielding values which we slice at the right point using `islice`.

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/IQVectors.py:
```diff
@@ -106,6 +107,132 @@ def hkl_to_qvectors(self, hkls: np.array, unit_cell: UnitCell) -> np.ndarray:
+            np.max(self.qvectors_to_hkl(start_shape * (qmin), unit_cell))
+        )
+        range_end = np.ceil(np.max(hkl_sphere))
+        steps = np.arange(range_start, range_end + 0.1, 1)
```
Why `+0.1` with an integer step? Is that not a little unstable for different hkl shapes?

MBartkowiakSTFC replied:
I will remove this method, since we are not using it at the moment.

oerc0122 reviewed (commented):
I think you probably want https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.truncnorm.html throughout your random number generation.

ChiCheng45 commented:
I get an error when I try to "preview vector distribution" in the Actions tab. I see the following error message.

```
2025-11-25 13:16:35,131 - ERROR - process[36072] - main 36 - EXCEPTION:
<class 'TypeError'>
'configurator' is an unknown keyword argument
<traceback object at 0x000001AA209AC1C0>
Traceback (most recent call last):
  File "C:\Users\...\PycharmProjects\MDANSE\MDANSE_GUI\Src\MDANSE_GUI\InputWidgets\QVectorsWidget.py", line 376, in helper_dialog
    self.helper = self.create_helper()
                  ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\...\PycharmProjects\MDANSE\MDANSE_GUI\Src\MDANSE_GUI\InputWidgets\QVectorsWidget.py", line 367, in create_helper
    dialog_instance = VectorViewer(self._base, configurator=self._configurator)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\...\PycharmProjects\MDANSE\MDANSE_GUI\Src\MDANSE_GUI\InputWidgets\QVectorsWidget.py", line 271, in __init__
    self.shell_panel_3D = ShellPanel(
                          ^^^^^^^^^^^
  File "C:\Users\...\PycharmProjects\MDANSE\MDANSE_GUI\Src\MDANSE_GUI\InputWidgets\QVectorsWidget.py", line 158, in __init__
    super().__init__(*args, **kwargs)
TypeError: 'configurator' is an unknown keyword argument
```

MBartkowiakSTFC commented:
> I get an error when I try to "preview vector distribution" in the Actions tab. I see the following error message.
> 

This should be fixed now.

I think that circular vectors don't work correctly at the moment. I'll let you know when that's done.

MBartkowiakSTFC commented:
Actually, circular vectors are probably OK, but the 3D scatter plot resets axes scaling in such a way that a circle that should be in a plane appears to be inclined out of plane.

ChiCheng45 commented:
The spherical plotting looking good. Here is what I see with spericalqvectors and 1000 vectors. 

<img width="1050" height="525" alt="image" src="https://github.com/user-attachments/assets/37cd7b80-d6e9-4664-a0b7-c392b1b8c8ce" />

The distribution for both polar and azimuthal angles looks right. To avoid the users getting confused when looking at the polar angle it might be a good idea to normalise these by $1/\sin(\phi)$, $2/ \sin(\phi)$ or something similar. See eq.5 in https://mathworld.wolfram.com/SpherePointPicking.html $0.5 \sin(\phi)$ is the distribution in the polar coordinate.

ChiCheng45 commented:
Do the vector plots have their weights applied before binning in the histograms? I'm plotting the sphericallattiveqvectors of a cubic cell and I see the following for one of the shells.

<img width="1002" height="530" alt="image" src="https://github.com/user-attachments/assets/b3b108f4-90b2-4f7d-8681-787cab06b560" />

<img width="507" height="262" alt="image" src="https://github.com/user-attachments/assets/e5a4d724-2c20-4a6a-96a1-a9816c0f9936" />

There are fewer lattive q-vectors for polar angles around $\pi / 2$, which makes sense. Hopefully, the weights should weight the vectors around  $\pi / 2$ more.

MBartkowiakSTFC commented:
> Do the vector plots have their weights applied before binning in the histograms? I'm plotting the sphericallattiveqvectors of a cubic cell and I see the following for one of the shells.
> 

I don't think that weights were being applied here. Good point, I'll fix it.

MBartkowiakSTFC commented:
> The distribution for both polar and azimuthal angles looks right. To avoid the users getting confused when looking at the polar angle it might be a good idea to normalise these by 1 / sin ⁡ ( ϕ ) , 2 / sin ⁡ ( ϕ ) or something similar. See eq.5 in https://mathworld.wolfram.com/SpherePointPicking.html 0.5 sin ⁡ ( ϕ ) is the distribution in the polar coordinate.

I am happy to normalise the angles, but won't it cause problems for other, non-spherical vector generators?  Then again, maybe the angle distribution is not even useful for most of other vector generators. Do you think this normalisation should be applied always, or just to the spherical ones? I'm not sure what's more confusing in the long run.

ChiCheng45 commented:
> > The distribution for both polar and azimuthal angles looks right. To avoid the users getting confused when looking at the polar angle it might be a good idea to normalise these by 1 / sin ⁡ ( ϕ ) , 2 / sin ⁡ ( ϕ ) or something similar. See eq.5 in https://mathworld.wolfram.com/SpherePointPicking.html 0.5 sin ⁡ ( ϕ ) is the distribution in the polar coordinate.
> 
> I am happy to normalise the angles, but won't it cause problems for other, non-spherical vector generators? Then again, maybe the angle distribution is not even useful for most of other vector generators. Do you think this normalisation should be applied always, or just to the spherical ones? I'm not sure what's more confusing in the long run.

I think it makes sense to have it always applied. I think if we have a distribution like $\rho(r, \theta, \phi)$, which could be the distribution or density of q-vectors or something, then the count between angles $x$ and $x+C$ would be
 
```math
N(x + C) - N(x) = \int_{x}^{x+ C} \int_{r_0}^{r_1} \int_{0}^{2\pi} \rho(r, \theta, \phi) r^2 \sin(\theta) d\theta dr d\phi = \int_{x}^{x + C} \tilde{\rho}(\theta)  \sin(\theta)  d\theta
```

if we make C small then

```math
N(x + C) - N(x) \approx  \frac{d N}{d x} C
```
and 
```math
\int_{x}^{x+ C} \tilde{\rho}(\theta)  \sin(\theta)  d\theta \approx \tilde{\rho}(x)  \sin(x) C
```
so that the function we are approximating with our histogram is
```math
\frac{d N}{d x} = \tilde{\rho}(x)   \sin(x)
```
so I think its better to normalise so we plot a histogram which approximates $\tilde{\rho}(x)$.

ChiCheng45 commented on MDANSE/Src/MDANSE/Framework/Jobs/CurrentCorrelationFunction.py:
```diff
@@ -451,7 +453,8 @@ def run_step(self, index: int):
                     curr = np.einsum(
                         "ik,ij->ikj",
                         veloc,
-                        np.exp(1j * np.dot(coords, qVectors)),
+                        np.exp(1j * np.dot(coords, qVectors))
+                        * np.sqrt(qvec_weights[None, :]),
                     )
```
Looks correct. No need for the new axis, could also do this.
```suggestion
                    curr = np.einsum(
                        "ik,ij,j->ikj",
                        veloc,
                        np.exp(1j * np.dot(coords, qVectors)),
                        np.sqrt(qvec_weights),
                    )
```

ChiCheng45 commented on MDANSE/Src/MDANSE/Framework/Jobs/CurrentCorrelationFunction.py:
```diff
@@ -438,7 +440,7 @@ def run_step(self, index: int):
                     curr = np.einsum(
                         "ik,ij->ikj",
                         veloc,
-                        np.exp(1j * temp_dotprod),
+                        np.exp(1j * temp_dotprod) * np.sqrt(qvec_weights[None, :]),
                     )
```
Here also, no need for new axis could also do this.

```suggestion
                    curr = np.einsum(
                        "ik,ij,j->ikj",
                        veloc,
                        np.exp(1j * temp_dotprod),
                        np.sqrt(qvec_weights),
                    )
```

MBartkowiakSTFC committed:
Respond to review comments

MBartkowiakSTFC commented:
It seems that the main branch generates less shells than this branch, but they are better defined. Here are the statistics for the NaF trajectory (cubic, 21 Angstrom) with q-range reproducing the experiment.

Protos branch:
<img width="4884" height="2660" alt="protos_lattice_vector_stats" src="https://github.com/user-attachments/assets/3dd1494a-1f73-4563-a188-af74e9085188" />

This branch:
<img width="4884" height="2660" alt="new_lattice_vector_stats" src="https://github.com/user-attachments/assets/6e095188-c2bb-4c79-965d-5d142764098e" />

I tried running DISF with the same parameters using `SphericalQVectors` and `SphericalLatticeQVectors`. I had to use the third shell, since protos would not generate the first two.

Protos comparison (lattice vectors vs. all vectors):
<img width="4884" height="2540" alt="protos_lattice_vs_nolattice_q0 3" src="https://github.com/user-attachments/assets/15338913-bfd3-44ab-9dbb-4bcb06178bc5" />

This branch:
<img width="4884" height="2540" alt="new_lattice_vs_nolattice_q0 3" src="https://github.com/user-attachments/assets/d16d4c75-5258-4b52-8e69-497a9199c780" />

I guess that the next step will be to compare which result is the closest to the experiment.

ChiCheng45 commented:
I think for the NaF case, protos would be better because it's a liquid system. It's already isotropic, so we won't get any benefit from the better distribution in the angular part. I suppose we need to figure out what the best balance is between a better angular distribution or a more well-defined width in |q|.

MBartkowiakSTFC commented:
I tried to make a quick comparison of the experimental data to the calculation, but it probably would have been more meaningful with a better trajectory. Also, I subtracted a constant flat background from the experimental data, and normalised all the results by dividing all points by S(q,0). This is the comparison of DISF results to the experiment, but there are probably too many assumptions that would need to be verified before this comparison can be used for anything.
<img width="1069" height="734" alt="comparison_plot" src="https://github.com/user-attachments/assets/691481c3-749c-45b7-b443-d584616e405e" />

My current perspective is that:
1. The new sampling of vectors with different weights is a good idea and it should stay in the code.
2. The new relaxed attitude towards the |q| values is too big a change in behaviour to be introduced at this stage, and I would prefer to keep the sharp limits on (q_min, q_max) in each shell. That is, the users will still be able to make broad, smeared and overlapping shells by increasing the width parameter, but setting it to a low value should produce narrow shells.

I will make changes to this PR to make q_width work in a more restrictive way and will then ask you to check it again. Further changes to q-vector generation can be added after the first release.

MBartkowiakSTFC committed:
Stop removing 0 vector from results

MBartkowiakSTFC committed:
Ensure even spacing of q-dependent result arrays

MBartkowiakSTFC committed:
Remove supercell input from dispersion lattice qvectors

MBartkowiakSTFC committed:
Update unit test results for CurrentCorrelationFunction

MBartkowiakSTFC commented:
The current procedure of vector generation is:
1. Generate vectors randomly in a specified shape. Typically, we use uniform distribution in angles and normal distribution around the target |q|.
2. Snap/round the vectors to nearest lattice vectors. Weights are determined based on the number of times each vector was repeated in the output.
3. Remove the vectors outside of the |q| range for each shell. This way q_width can be used to control the shell width.
4. Mark the shells that were meant to be created but ended up empty. The calculation results for these shells will be `NaN`. This is to make sure that the heatmap plots map the |q| values correctly, since it is assumed that the data points are on a uniform grid.

Other changes:
1. Plotter has been adjusted to handle `NaN` values. For the moment, the areas of heatmaps with NaN values are white.
2. Minor adjustments to plotter layout to reduce the overlap between subplots.
3. Additional corrections to `DispersionLatticeQVectors`. At the moment these only accept integer start and step values. Also, the `supercell` input has been removed. (In the end, if we want it at all, then we probably want it for all generators.)
4. Context menu in Plot Creator enables "Vector statistics" plot if you right-click a file that contains vector. To see the vector positions, you have to right-click a specific shell. It will only work for shells that are not empty.

MBartkowiakSTFC committed:
Update CCF results in test_grouping

ChiCheng45 reviewed (commented):
I have found one bug which occurs when you load a trajectory, select an instrument in the actions tab and press preview vector distribution. An error is seen in the logs.

```
2025-12-09 12:21:13,271 - ERROR - process[25588] - Vectors3D 62 - PlottingContext can't plot to None
2025-12-09 12:21:13,271 - ERROR - process[25588] - main 36 - EXCEPTION: <class 'AttributeError'> 'NoneType' object has no attribute 'add_subplot' <traceback object at 0x000002376C950A00>
```

MBartkowiakSTFC committed:
Merge branch 'protos' of https://github.com/ISISNeutronMuon/MDANSE into maciej/q-vector-generation

MBartkowiakSTFC committed:
Fix plotting problems and Single plot layout engine

MBartkowiakSTFC commented:
> I have found one bug which occurs when you load a trajectory, select an instrument in the actions tab and press preview vector distribution. An error is seen in the logs.
> 
> ```
> 2025-12-09 12:21:13,271 - ERROR - process[25588] - Vectors3D 62 - PlottingContext can't plot to None
> 2025-12-09 12:21:13,271 - ERROR - process[25588] - main 36 - EXCEPTION: <class 'AttributeError'> 'NoneType' object has no attribute 'add_subplot' <traceback object at 0x000002376C950A00>
> ```

This should work now.

While fixing this one, I noticed that the new settings of the matplotlib layout engine made "Single" plots look weird if there were too many curves in them. I tried to fix it by setting the figure's layout engine to`"none"` in the Single plotter.

Also, I added a fix to make the "Vectors" and "Vectors3D" plotters not selectable by the user. This way we can still use them in the code, but the user will not be able to try (and fail) to visualise normal datasets using those.

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Jobs/DynamicCoherentStructureFactor.py:
```diff
@@ -324,19 +331,24 @@ def run_step(self, index: int) -> tuple[int, dict[str, np.ndarray] | None]:
+                ).T[0] / (n_configs * norm)
                 self._outputData[f"dcsf/f(q,t)/{pair_str}"][index, :] += corr.real
+        else:
+            for pair_str, (_, _) in self.labels:
```
```suggestion
            for pair_str, _ in self.labels:
```
Don't really need to unpack if they're both ignored.

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/CircularLatticeQVectors.py:
```diff
@@ -49,78 +55,53 @@ class CircularLatticeQVectors(LatticeQVectors):
+            self._configuration["q_vectors"][q]["weights"] = weights
+            self._configuration["q_vectors"][q]["n_q_vectors"] = np.sum(weights)
+            self._configuration["q_vectors"][q]["q"] = q
+            self._configuration["q_vectors"][q]["hkls"] = lattice_hkl_vectors
```
Might be an opportunity to reduce repetition here.


```suggestion
            self._configuration["q_vectors"][q] = {
                "q_vectors": self.hkl_to_qvectors(lattice_hkl_vectors, self._unit_cell),
                "weights": weights,
                "n_q_vectors": np.sum(weights),
                "q": q,
                "hkls": lattice_hkl_vectors,
            }
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/CircularQVectors.py:
```diff
@@ -18,16 +18,84 @@
+    except ValueError:
+        return None
+    else:
+        return rotation.as_matrix()
```
No need for else after return


```suggestion
    return rotation.as_matrix()
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/CircularQVectors.py:
```diff
@@ -18,16 +18,84 @@
+
+def circle_rotation_matrix(
+    target_circle_axis: npt.NDArray[float],
+) -> npt.NDArray[float]:
```
```suggestion
) -> npt.NDArray[float] | None:
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/CircularQVectors.py:
```diff
@@ -43,47 +111,38 @@ class CircularQVectors(IQVectors):
+            self._configuration["q_vectors"][q]["q_vectors"] = q_vectors
+            self._configuration["q_vectors"][q]["n_q_vectors"] = nvecs_per_shell
+            self._configuration["q_vectors"][q]["weights"] = np.ones(nvecs_per_shell)
             self._configuration["q_vectors"][q]["q"] = q
```
```suggestion
            self._configuration["q_vectors"][q] = {
                        "q_vectors": q_vectors,
                        "n_q_vectors": nvecs_per_shell,
                        "weights": np.ones(nvecs_per_shell),
                        "q": q,
            }
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/DispersionLatticeQVectors.py:
```diff
@@ -40,32 +45,39 @@ class DispersionLatticeQVectors(LatticeQVectors):
+        dists = np.linalg.norm(q_vectors, axis=0)
+        sign_array = np.sign(np.dot(step_vector, hkl_vectors))
+        keyvals = dists.copy()
+        keyvals[np.where(sign_array < 0)] *= -1
```
Would this not be simpler (plus no where [O(N)] or intermediate array and contiguous memory) as:

```suggestion
        keyvals *= sign_array
```

MBartkowiakSTFC replied:
While this would be simpler, it would also replace the any vector orthogonal to the step direction with a 0.
If I start from (2,0,-10) along (0,0,1), the code without the `where` will generate a sequence with ..., (2,0,-1), (0,0,0), (2,0,1), ...

oerc0122 replied:
What about
```python
        keyvals[sign_array < 0] *= -1
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/DispersionLatticeQVectors.py:
```diff
@@ -40,32 +45,39 @@ class DispersionLatticeQVectors(LatticeQVectors):
-            self._configuration["q_vectors"][v]["hkls"] = hkls[:, i][:, np.newaxis]
+            self._configuration["q_vectors"][v]["hkls"] = hkl_vectors[:, i][
+                :, np.newaxis
+            ]
```
```suggestion
            self._configuration["q_vectors"][v] = {
                "q_vectors": q_vectors[:, i][:, np.newaxis],
                "n_q_vectors": weights[i],
                "weights": weights[i : i + 1],
                "q": v,
                "hkls": hkl_vectors[:, i][:, np.newaxis],
            }

```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/IQVectors.py:
```diff
@@ -29,16 +30,28 @@
     from MDANSE.Framework.OutputVariables.IOutputVariable import OutputData
     from MDANSE.MolecularDynamics.UnitCell import UnitCell
 
+GAUSS_WIDTH_FACTOR = 2 * 2.35482  # 2*sqrt(2*ln(2))
```
May as well go whole hog on the constant:

```suggestion
GAUSS_WIDTH_FACTOR = 4.70964  # 2*sqrt(2*ln(2))
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/IQVectors.py:
```diff
@@ -106,6 +119,53 @@ def hkl_to_qvectors(self, hkls: np.array, unit_cell: UnitCell) -> np.ndarray:
+    def remove_zero_vector(
+        self, vectors: npt.NDArray[float], weights: npt.NDArray[float] | None = None
+    ):
+        zero_vector_index = np.where(np.linalg.norm(vectors, axis=0) < 1e-15)[0]
```
Boolean arrays can be used as masks, you don't always need to convert to indices with where.

```python
import numpy as np

x = np.random.rand(10, 3)
print(x)
print(np.linalg.norm(x, axis=1))

y = x[np.linalg.norm(x, axis=1) < 1.]
print(y)
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/LinearLatticeQVectors.py:
```diff
@@ -55,64 +53,48 @@ class LinearLatticeQVectors(LatticeQVectors):
+            self._configuration["q_vectors"][q]["weights"] = weights
+            self._configuration["q_vectors"][q]["n_q_vectors"] = np.sum(weights)
+            self._configuration["q_vectors"][q]["q"] = q
+            self._configuration["q_vectors"][q]["hkls"] = lattice_hkl_vectors
```
```suggestion
            self._configuration["q_vectors"][q] = {
                "q_vectors": self.hkl_to_qvectors(lattice_hkl_vectors, self._unit_cell),
                "weights": weights,
                "n_q_vectors": np.sum(weights),
                "q": q,
                "hkls": lattice_hkl_vectors,
            }

```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/LinearQVectors.py:
```diff
@@ -55,29 +64,31 @@ def _generate(self):
-            self._configuration["q_vectors"][q]["n_q_vectors"] = nVectors
+            self._configuration["q_vectors"][q]["q_vectors"] = q_vectors
+            self._configuration["q_vectors"][q]["n_q_vectors"] = nvecs_per_shell
+            self._configuration["q_vectors"][q]["weights"] = np.ones(nvecs_per_shell)
```
```suggestion
            self._configuration["q_vectors"][q] = {
                "q_vectors": q_vectors,
                "n_q_vectors": nvecs_per_shell,
                "weights": np.ones(nvecs_per_shell),
            }
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/SphericalLatticeQVectors.py:
```diff
@@ -55,61 +54,40 @@ class SphericalLatticeQVectors(LatticeQVectors):
+            self._configuration["q_vectors"][q]["n_q_vectors"] = np.sum(weights)
+            self._configuration["q_vectors"][q]["weights"] = weights
+            self._configuration["q_vectors"][q]["q"] = q
+            self._configuration["q_vectors"][q]["hkls"] = lattice_hkl_vectors
```
```suggestion
            self._configuration["q_vectors"][q] = {
                "q_vectors": self.hkl_to_qvectors(lattice_hkl_vectors, self._unit_cell),
                "n_q_vectors": np.sum(weights),
                "weights": weights,
                "q": q,
                "hkls": lattice_hkl_vectors,
            }
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/SphericalQVectors.py:
```diff
@@ -53,25 +70,27 @@ def _generate(self):
+            self._configuration["q_vectors"][q]["q_vectors"] = q_vectors
+            self._configuration["q_vectors"][q]["n_q_vectors"] = nvecs_per_shell
+            self._configuration["q_vectors"][q]["weights"] = np.ones(nvecs_per_shell)
             self._configuration["q_vectors"][q]["q"] = q
```
```suggestion
            self._configuration["q_vectors"][q] = {
                "q_vectors": q_vectors,
                "n_q_vectors": nvecs_per_shell,
                "weights": np.ones(nvecs_per_shell),
                "q": q,
            }
```

oerc0122 commented on MDANSE/Tests/UnitTests/Analysis/test_scattering.py:
```diff
@@ -119,7 +119,7 @@ def test_dcsf(generate_benchmarks, tmp_path, traj_info, qvector_grid):
         scale_result=True,
         scale_benchmark=True,
-        atol=1e-8,
+        atol=1e-6,
```
Should this worry me?

MBartkowiakSTFC replied:
Probably not.

The other tests from this file are already using this tolerance. This was the only one left with more strict settings. At some point I reached a problem with three obvious solutions:
1) Make the test fail on Linux and Windows, but pass on Mac,
2) Make the test pass on Linux and Windows, but fail on Mac,
3) Change the `atol`.

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/InputWidgets/QVectorsWidget.py:
```diff
@@ -21,19 +21,43 @@
+        case 3:
+            return "rd"
+        case _:
+            return "th"
```
Unless we expect `-N`th values.
```suggestion
    match index % 10:
        case 1:
            return "st"
        case 2:
            return "nd"
        case 3:
            return "rd"
        case _:
            return "th"
```

MBartkowiakSTFC replied:
I think I messed this function up anyway, because we should have:
1st
**11th**
21st
31st
...

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/InputWidgets/QVectorsWidget.py:
```diff
@@ -118,6 +136,101 @@ def parse_vtype(self, vtype: str, value: str, vname: str):
+        current_shell = self.shell_selector.value()
+        self.shell_selector.setMaximum(number_of_shells - 1)
+        if current_shell >= number_of_shells:
+            current_shell = number_of_shells - 1
```
```suggestion
        current_shell = min(self.shell_selector.value(), number_of_shells - 1)
        self.shell_selector.setMaximum(number_of_shells - 1)
```

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/InputWidgets/QVectorsWidget.py:
```diff
@@ -118,6 +136,101 @@ def parse_vtype(self, vtype: str, value: str, vname: str):
+        if update_limit:
+            self.set_upper_limit(len(vec_dict))
+        try:
+            vec_key = list(vec_dict.keys())[shell_index]
```
```python
from more_itertools import nth
...

            vec_key = nth(vec_dict, shell_index)
```

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/InputWidgets/QVectorsWidget.py:
```diff
@@ -174,7 +302,11 @@ def __init__(self, *args, **kwargs):
+        self._selector.addItems(
+            filter(
+                lambda x: x not in ("LatticeQVectors",), IQVectors.indirect_subclasses()
+            )
```
```suggestion
            (
                x for x in IQVectors.indirect_subclasses() if x != "LatticeQVectors"
            )
```

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/InputWidgets/QVectorsWidget.py:
```diff
@@ -239,10 +371,9 @@ def create_helper(
-        dialog_instance.plot_widget._sliderpack.setVisible(False)
+        dialog_instance = VectorViewer(self._base, configurator=self._configurator)
+        for shell_panel in (dialog_instance.shell_panel_3D,):
+            self.new_shell_number.connect(shell_panel.set_upper_limit)
```
Any reason for a loop over one element?

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Plotters/Heatmap.py:
```diff
@@ -97,7 +97,9 @@ def change_normalisation(self, new_value: dict[str, Any]):
-            results = [np.percentile(new_data, perc) for perc in percentiles]
+            results = [
+                np.percentile(np.nan_to_num(new_data), perc) for perc in percentiles
+            ]
```
`perc` can be an array.
```python
>>> import numpy as np
>>> x = np.random.rand(100,)
>>> perc = np.linspace(0, 100, 21)
>>> np.allclose([np.percentile(x, p) for p in perc], np.percentile(x, perc))
True
```

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Plotters/Plotter.py:
```diff
@@ -314,10 +320,18 @@ def plot_blank(self, *, draw_cross: bool = True):
-        axes.set_xlabel(
-            "If you expected a plot, please check the settings you changed last."
+        label_text = (
+            override_label
+            if override_label
+            else "If you expected a plot, please check the settings you changed last."
+        )
```
Can be:
```suggestion
        label_text = override_label or "If you expected a plot, please check the settings you changed last."
```
if clearer.

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Views/PlotDataView.py:
```diff
@@ -53,16 +54,181 @@ def shell_to_modq(shell_index: int, parent: h5py.Dataset) -> npt.NDArray[float]:
+    tuple[SingleDataset, SingleDataset]
+        Datasets of polar and azimuthal angles.
+    """
+    inplane_r = np.sqrt(q_array[0, :] ** 2 + q_array[1, :] ** 2)
```
```suggestion
    inplane_r = np.linalg.norm(q_array[:2, :], axis=1)
```

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Views/PlotDataView.py:
```diff
@@ -53,16 +54,181 @@ def shell_to_modq(shell_index: int, parent: h5py.Dataset) -> npt.NDArray[float]:
+        ),
+        (azimuthal_angles, "Azimuthal angle", r"$\phi$", False),
+    ):
+        rounding_precision = 3
```
Could be constant outside loop

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Views/PlotDataView.py:
```diff
@@ -72,41 +238,81 @@ def convert_vectors_to_datasets(
+                    sum(parent_dset[f"shell_{shell_index}/weights"][:])
+                    for shell_index in valid_shells
                 ]
             )
```
By defining a generator within this function (so that `valid_shells` and `parent_dset` are captured in the closure):
```python
        def q_data(elem) -> Generator:
            for index in valid_shells:
                yield parent_dset[f"shell_{index}/{elem}"]
```
, a lot of this could be reduced:

```suggestion
        if all("weights" in dat for dat in q_data("")):
            vec_weights = [
                dat[:] for dat in q_data("weights")
            ]
            unique_vectors = np.array(
                [
                    len(dat) for dat in q_data("weights")
                ]
            )
            all_vectors = np.array(
[
                    sum(dat[:]) for dat in q_data("weights")
]
)
```

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Views/PlotDataView.py:
```diff
@@ -72,41 +238,81 @@ def convert_vectors_to_datasets(
-        qvals = np.array(source["shells"])
-        nshells = len(qvals)
+        qvals = np.array([float(x) for x in source["q_vectors"]])
+        valid_shells = [
```
Can do something similar to the generator above

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Visualisers/InstrumentInfo.py:
```diff
@@ -115,12 +115,12 @@ def sanitize_numbers(self):
         return results
 
-    def create_q_vector_params(self, sample_configuration=None):
+    def create_q_vector_params(self, sample_cell=None):
```
Typing?

MBartkowiakSTFC committed:
Add changes from code review

MBartkowiakSTFC committed:
Add changes from code review

MBartkowiakSTFC committed:
Merge branch 'protos' of https://github.com/ISISNeutronMuon/MDANSE into maciej/q-vector-generation

MBartkowiakSTFC committed:
Implement changes from review comments

ChiCheng45 commented on MDANSE/Src/MDANSE/Framework/QVectors/SphericalQVectors.py:
```diff
@@ -19,18 +19,35 @@
+            radii * np.sin(theta) * np.sin(phi),
+            radii * np.cos(theta),
+        )
+    )
```
This problem also exists on protos, but in a different form. Though the effect might be small for the widths that are typically used.

To uniformly sample in a sphere volume, you need to sample the angles correctly as you have done here, but you also sample in radii following this

```python
radii = r_max * random.uniform()**(1/3)
```
where `r_max` is the radius of the sphere. See, for example, https://math.stackexchange.com/questions/87230/picking-random-points-in-the-volume-of-sphere-with-uniform-probability. The `random.uniform()**(1/3)` is required because for small values of the radius, the smaller the shell volumes are. You need to sample larger radii more.

I think if you use `all_radii = truncated_normal_distribution(n_vecs, qmin, qmax, q_width, q)` then the sampling will be skewed towards small radii, though I think the effect is small.

ChiCheng45 replied:
Here is the issue in protos. 

<img width="903" height="611" alt="image" src="https://github.com/user-attachments/assets/3089ad77-9c4c-416d-a136-64b441f8d562" />

but this is not uniform; the plot is misleading.

Here is a code that demonstrates the correct distribution,

```python
import numpy as np
import matplotlib.pyplot as plt

xyz = 4 * np.random.rand(100000, 3) - 2

r = np.linalg.norm(xyz, axis=1)
r = r[np.logical_and(r > 1, r < 2)]

bins = np.arange(1, 2.1, 0.1)

plt.hist(r, bins=bins, edgecolor='black')
plt.xlabel('Value')
plt.ylabel('Count')
plt.show()
```` 

I generate a uniform distribution of points from -2 to 2 in x, y, and z. Then calculate the r and then take values from 1 to 2 and plot them in a histogram. I get the following.

<img width="440" height="337" alt="image" src="https://github.com/user-attachments/assets/e9cc7aa1-d668-4bff-96a9-ac4f49ee72b7" />

Basically, when you bin in `r`, the bin volumes are larger, so you get a larger count.

MBartkowiakSTFC replied:
You are right that this approach needs to be taken to generate vectors that give us equal probability of sampling each point in space. That is, this would give us uniform distribution of vectors in volume.

My question is: do we actually want the vectors to be uniformly distributed in volume?

Uniform distribution of _angles_ is important, as this is an approximation of random orientations of the system in a much larger, isotropic sample. As for the distribution of |q|, we introduce a finite width of the shell to account in some way for the finite resolution of the instrument. So far the intention behind my changes to the code was to make sure that the average |q| calculated from all the vectors in a shell is as close as possible to the nominal value of |q| for this shell.

In a neutron experiment, the reciprocal space volume is normally not uniformly sampled. Some normalisation is then applied to correct for the instrument geometry, wavelength spectrum and detector efficiency effects, but ultimately I am still not sure that our distribution of q vectors should be uniform in q space.

Another thing is that our generators assume that the width of each shell in q is the same, no matter what |q| is the centre of the shell. This I expect to be also unrealistic. Maybe for QENS this is close enough, but most likely this is not true in general.

ChiCheng45 replied:
OK sure, when widths are small the results will be the same anyway.

ChiCheng45 commented on MDANSE/Src/MDANSE/Framework/QVectors/IQVectors.py:
```diff
@@ -106,6 +119,50 @@ def hkl_to_qvectors(self, hkls: np.array, unit_cell: UnitCell) -> np.ndarray:
 
+    @classmethod
+    def lattice_vectors_with_weights(
+        self,
```
As its a classmethod should be `cls` instead of `self`. This occurs a few times in this file.

ChiCheng45 reviewed (commented):
Can you update the histogram so that it is centred at the shell centre?

<img width="1057" height="630" alt="image" src="https://github.com/user-attachments/assets/27dd33ba-d0c1-4121-bbf0-a930beaaf051" />

In the above I have one shell at 5 1/nm but the middle histogram bin is not centred at 5, It starts from around 5. This makes the distribution look asymmetric.

Similarly for

<img width="793" height="470" alt="image" src="https://github.com/user-attachments/assets/89d12052-ae27-496f-ae56-e300d7c9ae5a" />

can you centre the histograms so for the polar angle we have a bin in the middle around pi / 2 and for the azimuthal angle around 0?

Also, I think it might be worth it to always plot the polar and azimuthal angles fully from 0 to pi and -pi to pi.

<img width="784" height="475" alt="image" src="https://github.com/user-attachments/assets/8e9890e7-7ad8-435b-b040-609911935dd3" />

ChiCheng45 reviewed (commented):
<img width="1185" height="612" alt="image" src="https://github.com/user-attachments/assets/25d977c2-150d-4e57-89bd-a1a1c32f3afe" />

The histogram is empty when the q-vector settings are apparently valid (no warning in the GUI and I am not blocked from running the job). I used the DFT water trajectory to generate this.

<img width="1208" height="620" alt="image" src="https://github.com/user-attachments/assets/f502389c-e593-4874-9d31-38c37c787a57" />

Vector angle statistics says there are no vectors but according to the top left plot of the first figure, it looks like there are some.

Job runs fine, but the results are nan.

<img width="501" height="269" alt="image" src="https://github.com/user-attachments/assets/980269e3-2556-4b21-91f5-09d6c8bdeb36" />

Now, if I increase the width to 2, there are some q-vectors, and I get results, but I still see an empty histogram.

<img width="913" height="494" alt="image" src="https://github.com/user-attachments/assets/efb307a1-aa89-401f-90a1-cccf5e987803" />

ChiCheng45 commented:
Previously width=0 was possible in SphericalQVectors. Its not working anymore, is this intentional?

<img width="717" height="202" alt="image" src="https://github.com/user-attachments/assets/9337c01a-a4dc-4956-b0f9-5b188a40cd5d" />

MBartkowiakSTFC committed:
Improve q vector histogram bin limits

MBartkowiakSTFC committed:
Implement violin plots for q length distribution

MBartkowiakSTFC committed:
Scale violin plot width correctly on unit change

MBartkowiakSTFC committed:
Improve the binning of q-vector histograms

ChiCheng45 reviewed (commented):
Violin plots looks great. A few things I've noticed. With sphericalqvectors with a small width, a q-vectors shells (0, 1, 0.5), and 10000 n_vectors. 

<img width="403" height="275" alt="Screenshot 2026-01-08 at 10 29 04" src="https://github.com/user-attachments/assets/0c849c91-c36f-41f2-b6be-71e1363a6fba" />

So the first shell which is centered around zero has a maximum which is not zero?

Here I used a width of zero. Looks like there are still some issues when a width of zero is used?
<img width="406" height="268" alt="Screenshot 2026-01-08 at 10 29 36" src="https://github.com/user-attachments/assets/810ed72f-0a3b-405f-9cb3-5980432d8ffc" />

MBartkowiakSTFC commented:
> 
> Job runs fine, but the results are nan.
> 

My current idea is to change the QVectorsWidget behaviour:
1. Produce an error if ALL the shells are empty.
2. Produce a warning if ANY shell is empty.
3. Normal behaviour if every shell contains at least one vector.

@oerc0122 These changes make the GUI unit tests fail, because the jobs using the QVectorsWidget now produce a warning. What do you think would be the cleanest way of making it work?

MBartkowiakSTFC commented:
> So the first shell which is centered around zero has a maximum which is not zero?

The first shell will contain some vectors with non-zero |Q|. Since it is impossible to introduce vectors with a negative |Q| to compensate for it, the distribution will not be centred on 0. I think we just have to accept it.

> 
> Here I used a width of zero. Looks like there are still some issues when a width of zero is used? 

The y scale shows that the differences are of the order of 1e-16. I guess this is the plot of our numerical precision in calculating the vector length.

MBartkowiakSTFC committed:
Fix failing tests, improve histogram limits

MBartkowiakSTFC committed:
Improve docstrings, implement linter suggestions

MBartkowiakSTFC commented:
I can see that after my changes the errors in converter widgets don't get cleared, so I still need to fix this.

MBartkowiakSTFC committed:
Remove the valid_changed signal from input widgets

ChiCheng45 commented:
Looks great, I don't have any complaints about the viewer and the plots. Seems to do the job correctly. 
 
Just one thing is that it is a bit slow when switching the q-vector generator type. I had a quick look, and it looks like `def set_shell(...)` in `class ShellPanel` in `QVectorsWidget.py` is being called multiple times. Also seems to be called twice when switching the Vector shell in the Vector angle statistics tab. I think this is causing the slowdown.

oerc0122 commented:
> > Job runs fine, but the results are nan.
> 
> My current idea is to change the QVectorsWidget behaviour:
> 
> 1. Produce an error if ALL the shells are empty.
> 2. Produce a warning if ANY shell is empty.
> 3. Normal behaviour if every shell contains at least one vector.
> 
> @oerc0122 These changes make the GUI unit tests fail, because the jobs using the QVectorsWidget now produce a warning. What do you think would be the cleanest way of making it work?

If the warning's a desired behaviour, test it's produced (`with pytest.warns`) and use `warnings.simplefilter` to remove the offending warning from the checks in the relevant GUI test or make the GUI test more specific.

oerc0122 commented on MDANSE/Src/MDANSE/Framework/Configurators/QVectorsConfigurator.py:
```diff
@@ -121,3 +121,14 @@ def configure(self, value: tuple[str, dict[str, Any]]):
         self["generator"] = generator
         self.error_status = "OK"
+
+        if any(shell is None for shell in self["value"].values()):
```
```suggestion
        if None in self["value"].values():
```

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/DispersionLatticeQVectors.py:
```diff
@@ -60,25 +62,21 @@ def _generate(self):
         if self._status is not None:
             self._status.start(len(dists))
 
         self._configuration["q_vectors"] = collections.OrderedDict()
```
Just `{}` is the same since 3.7

oerc0122 commented on MDANSE/Src/MDANSE/Framework/QVectors/IQVectors.py:
```diff
@@ -30,10 +30,48 @@
+    npt.NDArray[float]
+        A truncated normal distribution of points within the specified limits.
+    """
+    if abs(width) < WIDTH_NONZERO_LIMIT:
```
Either works
```suggestion
    if math.isclose(width, 0, atol=WIDTH_NONZERO_LIMIT):
```

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Models/PlottingContext.py:
```diff
@@ -173,6 +174,7 @@ def _(
         yerror: npt.NDArray[float] | None = None,
         xerror: npt.NDArray[float] | None = None,
         optional_filename: str | None = None,
+        uneven_array: bool = False,
```
```suggestion
        optional_filename: str | None = None,
        *,
        uneven_array: bool = False,
```
Honestly, if this is able to be API breaking, I think a lot of these params want to be kwarg-only

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Models/PlottingContext.py:
```diff
@@ -190,21 +192,35 @@ def _(
+            If True, allow the data rows to be of different length.
         """
-
         self._filename = optional_filename if optional_filename else "no file"
```
Why pass in `None`? set default to `"no file"`

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Views/PlotDataView.py:
```diff
@@ -16,27 +16,144 @@
+    """
+    step_params = qvector_params.get("shells")
+    if step_params is None:
+        return None
```
Could be:
```suggestion
    if (step_params := qvector_params.get("shells")) is None:
        return None
```
Not sure if clearer or not.

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Views/PlotDataView.py:
```diff
@@ -16,27 +16,144 @@
+    npt.NDArray[float]
+        A 1D array of |q| histogram bin limits.
+    """
+    width = abs(width)
```
If `width` is `None` as implied by the signature, this will immediately fall over.

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Views/PlotDataView.py:
```diff
@@ -16,27 +16,144 @@
+    if np.isclose(start, end) and (width is None or abs(width) < WIDTH_NONZERO_LIMIT):
+        return np.array([start - 0.15, start - 0.05, start + 0.05, start + 0.15])
+
+    def get_bin_width(n_segments):
```
```suggestion
    def get_bin_width(n_segments: int) -> tuple[float, float]:
```

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Views/PlotDataView.py:
```diff
@@ -16,27 +16,144 @@
+        else:
+            bin_width = step_size / n_segments
+            peak_width = width
+        return bin_width, peak_width
```
Could use earlier returns
```suggestion
        if np.isclose(start, end):
            return width / n_segments, width

        if width is None or width <= step_size / 2:
            return step_size / n_segments, step_size

        return step_size / n_segments, width
```

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Views/PlotDataView.py:
```diff
@@ -16,27 +16,144 @@
+        )
+        if len(common_binning) == last_value:
+            break
+        last_value = len(common_binning)
```
I'm slightly confused on the naming and logic here. You seem to be using `last_value` to mean two things simultaneously. The previous value of the loop and the final value in the binning range. Not sure if this is sensible.

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Plotters/Vectors.py:
```diff
@@ -233,33 +219,55 @@ def plot(
+        for axindex, axes in enumerate(self._axes):
             if update_only:
+                plot_limits = self._backup_limits[axindex]
+                if axindex != 1:
```
Don't particularly like the logic here. I have a gut feeling it could be clearer, but I can't immediately see a nicer way to do it.

MBartkowiakSTFC committed:
Respond to review comments

MBartkowiakSTFC committed:
Remove most occurrences of collections.OrderedDict

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/InputWidgets/QVectorsWidget.py:
```diff
@@ -69,6 +71,24 @@ def numerator_suffix(index: int) -> str:
 
+def dicts_are_the_same(new_dict: dict[str, str], ref_dict: dict[str, str]) -> bool:
+    """Check if all the values in the new dict are the same in the reference dict."""
+    return all(new_dict.get(key) == value for key, value in ref_dict.items())
```
Does `new_dict == ref_dict` not work?

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Tabs/Views/PlotDataView.py:
```diff
@@ -95,27 +95,28 @@ def qvector_binning_general(
         start, end = end, start
-    if np.isclose(start, end) and (width is None or abs(width) < WIDTH_NONZERO_LIMIT):
+    if np.isclose(start, end) and (
+        width is None or math.isclose(width, 0, abs_tol=WIDTH_NONZERO_LIMIT)
```
Any reason for using `np.isclose` for the first and `math.isclose` for the second?

oerc0122 commented on MDANSE_GUI/Tests/UnitTests/test_qvector_widget.py:
```diff
@@ -0,0 +1,90 @@
+    """
+    window = QMainWindow()
+    jobname = "DynamicCoherentStructureFactor"
+    index = [job_index for job_index, job_name in enumerate(sorted(ENABLED_JOBS), 1) if job_name == jobname][0]
```
Lots of opportunity for confusion between `job_name` and `jobname`. `curr_job_name`?

MBartkowiakSTFC committed:
Respond to review comments

MBartkowiakSTFC commented:
After the changes, the vector preview dialog seems to update only when needed. That is, it will update the plots in the active tab, if the dialog is visible and the parameters have changed. This should be a sufficient solution to the problem of slow switching between different q vector generators.

As for the unit tests, I created a trajectory with a very large cell so that no warnings are produced during the GUI widget tests. Then we have a specific test for the `QVectorsWidget` where a normal trajectory is used. There we check that the generators that should produce a warning do produce it, while the others (i.e. most of them) don't.

Otherwise I made changes as suggested in code review.

oerc0122 commented on MDANSE_GUI/Tests/UnitTests/test_qvector_widget.py:
```diff
@@ -18,7 +18,11 @@
+    qvecs
+    for qvecs in IQVectors.indirect_subclasses()
+    if qvecs not in ("IQVectors", "LatticeQVectors")
+]
```
If order doesn't matter, can use sets for cleaner removal.
```suggestion
ENABLED_QVECTORS = set(IQVectors.indirect_subclasses()) - {"IQVectors", "LatticeQVectors"}
```

MBartkowiakSTFC committed:
Use set operations in test_qvector_widget.py

oerc0122 reviewed (approved):
If Chi's happy I think this is ready codewise
